### ISSUE TITLE
[Qustion] On the training the CNN networks

### ISSUE THREAD

**Issue by NeighborhoodCoding** [lidq92/VSFA#37]
Hi, I want to train the CNN instead of transfer learning and freeze it.

So I integrated CNN part and LSTM part to one class, also modified the train data module.

However, the dimension of CNN output was [timesteps x features x batches x 1]

and input of GRU was [batches x timesteps x features].

Because CNN output and GRU(batchfirst=True) input has difference in terms of dimension,

I used squeeze and unsqueeze and  swapaxes technique and it worked. (I'm not 100% sure)

Here is my code.

```
    def forward(self, pixel, video_length):
        for avg_pooling, model in enumerate(self.batch):
            pixel = model(pixel)
            if avg_pooling == 1:
                #print("D0_1 :", pixel.shape)  # timesteps 576(FILTERS) 1 1
                avg_pooling_2d = nn.functional.adaptive_avg_pool2d(pixel, 1) # timesteps 576(FILTERS) 1 1
                #print("D0_2 :", avg_pooling_2d.shape)
                #avg_pooling_2d = torch.swapaxes(avg_pooling_2d, 2, 0)
                #avg_pooling_2d = torch.swapaxes(avg_pooling_2d, 1, 2)
                #avg_pooling_2d = torch.squeeze(avg_pooling_2d, dim=3)
                #avg_pooling_2d = SA(avg_pooling_2d)
                #print("D1 :", avg_pooling_2d.shape)   #1 timesteps FILTERS
                input = self.ReLU(avg_pooling_2d)
                #print("D2 :", input.shape) #1 24 256
                outputs, _ = self.rnn(input, torch.zeros(1, 1, 64, device=device))
```

I used batch size to 1 to simplity...
But My train and validation loss is too high... and I think the swapaxes? maybe a reason for unsusseful training....?(loss backpropagation is errored cuz of swapaces...? I'm not sure)

Could you may update your code that may train the CNN instead of just load it??
< I may expect much better performance when it may succeess...^^;;; >
I made it but I think my code has un error in somewhat although it can run in python....
(I experienced CUDA memory out, so I used some timesteps like using every 8th frames, that helped resolve the CUDA memory out)

Thanks..

**Comment by NeighborhoodCoding** [comment]
Finally, I succeeded in bringing CNN and LSTM together, and I think I could train in correct way now.

First, CNN does not have a time dimension, so I add the time dimension with "for loop" in CNN(for number of frames) and then LSTM goes after the same as before.

The problem is that the longer the epoch, the lower the performance. In my opinion, CNN is already a well-trained neural network, but the lack of learning data makes it rather overfit, and I don't know the exact reason yet.

Oh, by the way, I'm thinking of replacing CNN with Koncept512. Once replaced with koncept512 instead of resnet, performance did not improve significantly. I thought koncept512 was a neural network that measures the mos of images, so I thought it was better to transfer learning than a neural network that classifies images, but I'm not sure why it hasn't improved much yet.

Now I'm going to use multiple CNNs as an ensemble. What I want is for the error of mae to be 0.2 in the validation set. Is it possible? My research senior is asking for 0.2 because he cannot accept a 0.3 mae error. He asked why AI has an error performance of 0.3 out of 5 even if AI is trained, and he said mae that higher 0.1 is not AI.

**Comment by lidq92** [comment]
@SungmanHong Good luck! This may be related to overfitting problem, and I would reply to you in another issue #38
